core/src/main/java/jenkins/util/io/PathRemover.java

```diff
@@ -0,0 +1,285 @@
+ * THE SOFTWARE.
+ */
+
+package jenkins.util.io;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import hudson.Functions;
+import hudson.Util;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import java.nio.file.attribute.PosixFileAttributes;
+import java.nio.file.attribute.PosixFilePermission;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@Restricted(NoExternalUse.class)
+public class PathRemover {
+
+    public static PathRemover newSimpleRemover() {
+        return new PathRemover(ignored -> false);
+    }
+
+    public static PathRemover newRemoverWithStrategy(@Nonnull RetryStrategy retryStrategy) {
+        return new PathRemover(retryStrategy);
+    }
+
+    public static PathRemover newRobustRemover(int maxRetries, boolean gcAfterFailedRemove, long waitBetweenRetries) {
+        return new PathRemover(new PausingGCRetryStrategy(maxRetries < 1 ? 1 : maxRetries, gcAfterFailedRemove, waitBetweenRetries));
+    }
+
+    private final RetryStrategy retryStrategy;
+
+    private PathRemover(@Nonnull RetryStrategy retryStrategy) {
+        this.retryStrategy = retryStrategy;
+    }
+
+    public void forceRemoveFile(@Nonnull Path path) throws IOException {
+        for (int retryAttempts = 0; ; retryAttempts++) {
+            Optional<IOException> maybeError = tryRemoveFile(path);
+            if (!maybeError.isPresent()) return;
+            if (retryStrategy.shouldRetry(retryAttempts)) continue;
+            IOException error = maybeError.get();
+            throw new IOException(retryStrategy.failureMessage(path, retryAttempts), error);
+        }
+    }
+
+    public void forceRemoveDirectoryContents(@Nonnull Path path) throws IOException {
+        for (int retryAttempt = 0; ; retryAttempt++) {
+            List<IOException> errors = tryRemoveDirectoryContents(path);
+            if (errors.isEmpty()) return;
+            if (retryStrategy.shouldRetry(retryAttempt)) continue;
+            throw new CompositeIOException(retryStrategy.failureMessage(path, retryAttempt), errors);
+        }
+    }
+
+    public void forceRemoveRecursive(@Nonnull Path path) throws IOException {
+        for (int retryAttempt = 0; ; retryAttempt++) {
+            List<IOException> errors = tryRemoveRecursive(path);
+            if (errors.isEmpty()) return;
+            if (retryStrategy.shouldRetry(retryAttempt)) continue;
+            throw new CompositeIOException(retryStrategy.failureMessage(path, retryAttempt), errors);
+        }
+    }
+
+    @Restricted(NoExternalUse.class)
+    @FunctionalInterface
+    public interface RetryStrategy {
```

Comment:
Perhaps you can use the class https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/util/Retrier.java It's a generic retry-n-times class that can save some code. You can see some usage examples at https://github.com/jenkinsci/jenkins/blob/master/core/src/test/java/hudson/util/RetrierTest.java

Reply:
Or even https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/model/AperiodicWork.java could be helpful

Reply:
In these three methods: _forceRemoveFile_, _forceRemoveDirectoryContents_ and _forceRemoveRecursive_ we are repeating a try-retry logic.

Reply:
Yes, I've been struggling to figure out a nice clear way to reduce that duplication. As a functional programmer, what I'd really prefer is a way to express these as monads (short-circuiting) and applicatives (error-collecting), though there's no good way to represent that in Java itself.